### DISCUSSION TITLE
feature: separate player and transcript shortcodes

### DISCUSSION THREAD

**Discussion by composerjk** [hyperaudio/wordpress-hyperaudio#130] [General]
@maboa: A little while back, I started a quick implementation of separate player and transcript shortcodes. So, one could place the player and transcripts in different spots; also, make it easier to have two transcripts attached to the same hyperaudio player. I'm in the process of moving this to live, so thought I would mention. Will try to clean up more in case you're interested in including something like it.

Here's an example with two synchronized transcript windows: transcript + chat. Transcripts are files in custom fields. Set up with something like:

Set up the **player**:
```
[hyperaudio_player id="haplayer1" width="100%" height="600px" transcript-height="1em" show-active="true" src='https://player.vimeo.com/video/{{dc:post:meta key="video_id"}}']
```

**Transcript 1** in one spot:
```
[hyperaudio_transcript id="transcript1" player_id="haplayer1" file="{{dc:acf:post_field field="transcript_file"}}" show_hyperaudio_production="false"]
```

Chat 1 (i.e., **transcript 2**) in another spot:
```
[hyperaudio_transcript id="chat1" player_id="haplayer1" file="{{dc:acf:post_field field="chat_file"}}"]
```
<img width="1226" alt="image" src="https://github.com/hyperaudio/wordpress-hyperaudio/assets/3038883/7d1d3bbd-c6be-4c99-b36e-ec19a4854b11">

The current code for the shortcodes is on our [player-transcript-shortcodes branch](https://github.com/asmac-org/wordpress-hyperaudio/tree/player-transcript-shortcodes). I also put together a [test-separate-shortcodes-release branch](https://github.com/asmac-org/wordpress-hyperaudio/tree/test-separate-shortcodes-release) to include recent PRs #123 and #124 I submitted since I'm moving this from staging to production. This still more cleanup to do, but I thought I'd share the progress and add a note. Will try to cleanup at some point and submit a PR when ready.

I also added an option to disable the *A Hyperaudio Production* mark with a `show_hyperaudio_production` attribute on the `hyperaudio_transcript` shortcode, so that one doesn't end up with it on every transcript.

I was just reminded that I had made a change (not included in the branches) to the default `scrollerOffset` from `0` to `-160` so that the highlighted portions of the transcripts are more centered. Will open an issue to mention adding support of specifying that value.

**Comment by maboa** [comment]
Thanks for all of this @composerjk – looks like an interesting development. I'll take a look in the coming days.

**Comment by maboa** [comment]
Hi @composerjk – apologies for the delay in getting back to you. First of all, thanks for looking into this. I'm a little hesitant to merge this because one of the key objectives with this plugin is to make it as easy as possible to use – I wouldn't want to lose the simplicity we already have, so I think any changes should maintain backward compatibility, the issue there would be the added complexity of the codebase. 

To help me evaluate – can I ask what you see as the use case for multiple transcripts connected to the same source?

Regarding the other issues `show_hyperaudio_production` and `scrollerOffset` I think we could proceed with those, although we should probably standardise a naming convention – camelCase or snake_case.

In any case I would kindly ask that each issue be separated into separate issues.

I'm going to convert this issue to a discussion so we can better go through the finer points. Hope that works for you.

**Reply by composerjk** [reply]
Sure, our **primary use case:**

Video of Zoom event. Transcript 1 is the actual transcript. Transcript 2 is the Zoom chat. Our events tend to have a lively chat with links provided for additional related resources along the way during 2–4 hour events. That's the example use case shown in the image and initial note.

With that, the chat transcript for a public video might only be visible to a member logged in. So, there are conditionals around Transcript 2, in this case. Also, the transcript sections only show if a transcript (or chat) is available. In our implementation, the transcript and chat are uploaded as files referenced by custom fields for the custom post type.

Separating the transcripts out also makes it easier to place them in different spots.

**Reply by composerjk** [reply]
> Regarding the other issues show_hyperaudio_production and scrollerOffset I think we could proceed with those, although we should probably standardise a naming convention – camelCase or snake_case.

Yeah, that code was thrown together while I was working on different projects, so definitely needs to be cleaned up.

Will clean up and open a separate issue for the `show_hyperaudio_production` one; already opened #129 for `scrollerOffset`.

**Reply by composerjk** [reply]
> I wouldn't want to lose the simplicity we already have, so I think any changes should maintain backward compatibility, the issue there would be the added complexity of the codebase.

Totally understand that. And since it was just in initial proof of concept and the code needed to be cleaned up, I didn't submit a pull request. Especially since I knew there was duplicate code that shouldn't be duplicated; that quick first pass of implementation.

The simplicity would remain as I was expecting that the initial `[hyperaudio ...]` shortcode would still exist and that the two new shortcodes would be in addition to it, for those who wanted to separate out the transcripts from the player. I don't think I removed any of the original functionality.

Will work on cleaning up the code as I get back to those pieces. We have something like 230 videos, at the moment, that use the new shortcodes, though I haven't yet uploaded the transcripts and Zoom chats.

**Reply by maboa** [reply]
Oh I see, that all makes sense. 👍🏼  Interesting use case. Thanks for answering so comprehensively. I'll take a good luck at the PRs whenever you're ready.

**Comment by composerjk** [comment]
I suppose another possible use case for multiple transcripts could be showing multiple translations simultaneously. But, that is not our use case.

**Reply by maboa** [reply]
Interesting yes. I can imagine this for multiple viewers looking at the same screen.
